criu/util.c

```diff
@@ -652,8 +652,8 @@ int cr_daemon(int nochdir, int noclose, int *keep_fd, int close_fd)
 		if (close_fd != -1)
 			close(close_fd);
 
-		if ((*keep_fd != -1) && (*keep_fd != 3)) {
-			fd = dup2(*keep_fd, 3);
+		if (*keep_fd != -1) {
+			fd = dup(*keep_fd);
```

Comment:
If one of 0,1, or 2 isn't used in this process, dup() will return it, and then we will overwrite it by /dev/null, so we can't use dup here, don't we?

Reply:
Thank you for pointing this out. Perhaps we could use `fcntl(F_DUPFD)` instead?

Reply:
Yes, we can. But I see another problem here. keep_fd is used to protect a file descriptor which is going to be used after cr_daemon. Here we have two such descriptros. Why do we want to handle only one of them?

Reply:
> keep_fd is used to protect a file descriptor which is going to be used after cr_daemon.

I don't understand why, though, after fork() the child will inherit a copy of all file descriptors, right? Why do we need to 'protect' `keep_fd`?

Reply:
Why do we need keep_fd?

Reply:
@avagin https://github.com/checkpoint-restore/criu/pull/681/commits/bfb73aecf032c1779190d88c5224b546772ff259 Would that work? Do we want a test for it?